In [1]:
import numpy as np

In [2]:
import pandas as pd

importing the cleaned data

In [3]:
data= pd.read_csv(r"C:\Users\G.Sreenivasulu\ABHI\loan_approval_project\dataset\01_cleaned.csv")

LOAN PERFORMANCE KPIs - tells how loan system is performing overall

In [4]:
total_applicants =len(data)

approved_count = (data['loan_status']=="Approved").sum()
rejected_count = (data['loan_status'] == "Rejected").sum()

approval_rate = (approved_count/total_applicants) * 100
rejection_rate = (rejected_count/total_applicants) * 100

print("The total applicants for the loan is :", total_applicants)
print("The no.of applicants approved for the loan is:", approved_count)
print("The no.of applicants rejected for the loan is:",rejected_count)
print("The approval rate for the loans is:", round(approval_rate,2),"%")
print("The rejection rate for the loans is:", round(rejection_rate,2),"%")

The total applicants for the loan is : 4269
The no.of applicants approved for the loan is: 2656
The no.of applicants rejected for the loan is: 1613
The approval rate for the loans is: 62.22 %
The rejection rate for the loans is: 37.78 %


In [5]:
# comparing the average loan amount for both approved and rejected

app_avg_amount = data[data['loan_status'] == "Approved"]['loan_amount'].mean()
rej_avg_amount = data[data['loan_status'] == "Rejected"] ['loan_amount'].mean()

print("The avg loan amount of the approved loans is:",round(app_avg_amount,2))
print("The avg loan amount of the rejected loans is:",round(rej_avg_amount,2))

The avg loan amount of the approved loans is: 15247251.51
The avg loan amount of the rejected loans is: 14946063.24


In [6]:
#comparing the average loan term for both approved and rejected

app_avg_term = data[data['loan_status'] == "Approved"]['loan_term'].mean()
rej_avg_term = data[data['loan_status'] == "Rejected"]['loan_term'].mean()

print("The avg loan term of approved loans is:", round(app_avg_term,2))
print("The avg loan term of rejected loans is:", round(rej_avg_term,2))

The avg loan term of approved loans is: 10.4
The avg loan term of rejected loans is: 11.73


The above code snippet gives us an important insight
The rejected loans have a higher average loan_term, this shows that longer loan term mean smaller EMIs which is a risky factor to the banks.
So, Rejected loans tend to ask for longer repayment periods.
Loan term is a significant risk factor

CREDIT RISK KPIs - tells if the borrower can repay the loan back

In [7]:
# average CIBIL score - shows the repayment history and financial discipline

app_avg_cib = data[data['loan_status'] == "Approved"]['cibil_score'].mean()
rej_avg_cib = data[data['loan_status'] == "Rejected"]['cibil_score'].mean()

print("The avg CIBIL score of approved loans is:", round(app_avg_cib,2))
print("The avg CIBIL score of rejected loans is:", round(rej_avg_cib,2))


The avg CIBIL score of approved loans is: 703.46
The avg CIBIL score of rejected loans is: 429.47


The above code snippet tells us that there is a huge difference between avg CIBIL score of approved and rejected loans. This tells us that the loan applicants with low CIBIL scores are rejected. So, credit history is also a significant loan approval deciding factor.

In [8]:
print("The minimum CIBIL score is:",min(data['cibil_score']))
print("The minimum CIBIL score is:",max(data['cibil_score']))

The minimum CIBIL score is: 300
The minimum CIBIL score is: 900


In [9]:
# Let us create the CIBIL scores into the following buckets to find the approval rate in each range

bins =[300,499,599,699,799,900]
labels = ['300-499','500-599','600-699','700-799','800-900']
#giving each row its respective bin
data['cibil_bucket'] = pd.cut(data['cibil_score'],bins=bins, labels=labels,right=True)
#right =True includes the upper part of the range
#creating the approval rates for each bucket
cibil_approval_rates= data.groupby('cibil_bucket')['loan_status'].apply(lambda x:(x=="Approved").mean() *100)
print(round(cibil_approval_rates,2))


cibil_bucket
300-499    10.44
500-599    52.25
600-699    99.43
700-799    99.46
800-900    99.57
Name: loan_status, dtype: float64


C:\Users\G.Sreenivasulu\AppData\Local\Temp\ipykernel_15100\3453067272.py:9: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  cibil_approval_rates= data.groupby('cibil_bucket')['loan_status'].apply(lambda x:(x=="Approved").mean() *100)


The above result shows that the loans with cibil score less than 499 have less chance of approval, loans between 500 and 600 have moderate chances of approval, where as loans with cibil score above 600 have a very high approval rate

INCOME AND ASSETS KPIs - show repayment and financial capacity

#AVG annual income(app vs rej)

app_avg_inc = data[data['loan_status'] == "Approved"]['income_annum'].mean()
rej_avg_inc = data[data['loan_status'] == "Rejected"]['income_annum'].mean()

print("The average annual income for approved loans is :", round(app_avg_inc,2))
print("The average annual income for rejected loans is :", round(rej_avg_inc,2))


The difference between avg annual income for approved and rejected loans is very little. It is not enough to determine how much it affects loan status

In [10]:
#so, we use the income-to-loan ratio calculated before
data['inc_loan']
avg_ratio = data.groupby('loan_status')['inc_loan'].mean()
print("The avg income to loan ratio is:\n",round(avg_ratio,4))

The avg income to loan ratio is:
 loan_status
Approved    0.3445
Rejected    0.3578
Name: inc_loan, dtype: float64


In the above result, approved income-to-loan ratio states that for approved loans,annual income is 34.45% of the loan amount and for rejected loans, the annual income is 35.78% of the loan amount

And the ratio is high for rejected loans than the approved ones. This shows that for the rejected loans, the income is high. But, they are still rejected shows that income is not the only factor but there are other factors that influence the loan_status

In [11]:
avg_total_assets_app = data[data['loan_status'] =="Approved"]['total_assets'].mean()
avg_total_assets_rej = data[data['loan_status'] =="Rejected"]['total_assets'].mean()

print("The avg of total assets of approved loans is:",round(avg_total_assets_app,2))
print("The avg of total assets of rejected loans is:", round(avg_total_assets_rej,2))

The avg of total assets of approved loans is: 32377296.69
The avg of total assets of rejected loans is: 32831122.13


In the above result, the avg of total assets of rejected loans is higher than that of approved loans, which tells us that even with high assets,some are rejected. To understand this more clearly we look into the distirbutions

In [12]:
#considering how assets affect the loan_status, using the total_assets calculated before

print("min:",min(data['total_assets']))
print("max:",max(data['total_assets']))

asset_bins = [0, 1000000, 5000000, 10000000, 25000000, 100000000]
asset_labels =['0-10L','10L-50L','50L-100L','100L-250L','250L-10C']
data['asset_bucket']= pd.cut(data['total_assets'],bins=asset_bins,labels=asset_labels,right=False)
asset_approval_rate = data.groupby('asset_bucket')['loan_status'].apply(lambda x:(x=="Approved").mean() *100)
asset_approval_rate
# cibil_approval_rates= data.groupby('cibil_bucket')['loan_status'].apply(lambda x:(x=="Approved").mean() *100)
# data

min: 400000
max: 90700000


C:\Users\G.Sreenivasulu\AppData\Local\Temp\ipykernel_15100\2124938501.py:9: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  asset_approval_rate = data.groupby('asset_bucket')['loan_status'].apply(lambda x:(x=="Approved").mean() *100)


asset_bucket
0-10L        38.461538
10L-50L      65.714286
50L-100L     60.285714
100L-250L    63.768116
250L-10C     61.597839
Name: loan_status, dtype: float64

From the above results, we can understand that till 50L, the approval increased sharply. But, from 50L, the approval rate lingered between 60 and 64, it indicates the dimnishing returns of very high asset values on approval.

So, the assets KPI shows that risk even for high asset values. It shows a non-linear relationship

In [13]:
#SUMMARY OF ALL THE KPIs

print("LOAN PERFORMANCE KPI")
print("The avg loan term of approved loans is:", round(app_avg_term,2))
print("The avg loan term of rejected loans is:", round(rej_avg_term,2))
print("HIGH LOAN TERM == REJECTED LOANS and less gap")

LOAN PERFORMANCE KPI
The avg loan term of approved loans is: 10.4
The avg loan term of rejected loans is: 11.73
HIGH LOAN TERM == REJECTED LOANS and less gap


In [14]:
print("CIBIL SCORE KPI")
print("The avg CIBIL score of approved loans is:", round(app_avg_cib,2))
print("The avg CIBIL score of rejected loans is:", round(rej_avg_cib,2))
print("HIGH CIBIL SCORE == APPROVED LOANS and huge gap")


CIBIL SCORE KPI
The avg CIBIL score of approved loans is: 703.46
The avg CIBIL score of rejected loans is: 429.47
HIGH CIBIL SCORE == APPROVED LOANS and huge gap


In [15]:
print("INCOME KPI")
print("The avg income to loan ratio is:\n",round(avg_ratio,4))
print("REJECTED LOANS RATIO > APPROVED LOANS RATIO & small gap")

INCOME KPI
The avg income to loan ratio is:
 loan_status
Approved    0.3445
Rejected    0.3578
Name: inc_loan, dtype: float64
REJECTED LOANS RATIO > APPROVED LOANS RATIO & small gap


In [16]:
print("ASSETS KPI")
print("The avg of total assets of approved loans is:",round(avg_total_assets_app,2))
print("The avg of total assets of rejected loans is:", round(avg_total_assets_rej,2))
print("REJECTED LOAN ASSETS > APPROVED LOAN ASSETS & small gap")

ASSETS KPI
The avg of total assets of approved loans is: 32377296.69
The avg of total assets of rejected loans is: 32831122.13
REJECTED LOAN ASSETS > APPROVED LOAN ASSETS & small gap


So, among all the KPIs, CIBIL score has the strongest influence on loan approval decisions compared to income and assets

In [18]:
data.to_csv(r"C:\Users\G.Sreenivasulu\ABHI\loan_approval_project\dataset\02_KPI_dataset.csv")